assets/test/channel_test.js

```diff
@@ -49,11 +49,15 @@ describe("constructor", () => {
     assert.equal(joinPush.timeout, 1234)
   })
 
+  it("throws if instantiating the same channel twice", () => {
+    socket.channel("topic", {one: "two"})
```

Comment:
Can't figure out why `socket.channel("topic", {one: "two"})` is throwing the error while 
```js
channel = socket.channel("topic", {one: "two"})
assert.equal(socket.channels.length, 1)

channel = socket.channel("topic", {two: "three"})
assert.equal(socket.channels.length, 1)
````

is not

Reply:
`socket.channel(topic)` pushes a new channel into `socket.channels` for the first time it's called and then if called twice, will just pick up the channel with that specific topic and return. We just return the current channel to prevent throwing an exception which we can avoid throwing.

Although if for some reason, either a race condition or a manual instantiation of a new channel happens (through `new Channel()` for instance), we need a safeguard to prevent that channel from being created as it would lead to double-bindings, double-joins, etc.

We could just raise an error if a channel is instantiated twice for all scenarios but I feel like returning the channel gracefully for the `socket.channel(topic)` use-case – which is probably 99% of the use cases – is more user-friendly and less prone to issues.

Reply:
Let me know your thoughts @achempion : )
Also, I sent a proposal to standardize this on the Phoenix mailing list.

Reply:
> We could just raise an error if a channel is instantiated twice for all scenarios but I feel like returning the channel gracefully ...

```js
it("throws if instantiating the same channel twice", () => {
```
Is it the case where we should return the old channel instead of throwing an error? 
Because of instantiating the same channel twice.

Reply:
Looks like the exception will raise only with some sort of race conditions, maybe we can mention that in test description to be more precise. Still can't understand how we can simulate race conditions with test environment :)

Reply:
> Looks like the exception will raise only with some sort of race conditions
That or if the person manually instantiates a new instance of channel through `new Channel` – which I am not even sure if it's applicable use, honestly : )

> Is it the case where we should return the old channel instead of throwing an error?
I think the error, in this case, is more idiomatic as it prevents misuse (e.g.: raising an error when the person is instantiating the same channel twice?)

Summarizing what I tried to accomplish:

1) if you're using `socket.channel('topic')` you will always get a channel being returned. If you call it once it will return a `new Channel()` and for the second time it will pick it up given a certain topic.

2) if for some reason a channel with the same topic is instantiated twice, it will raise an error. This is to prevent any possible race condition on the front end while instantiating the new channel.

3) the `join` race condition is already sorted by `joinedOnce` flag.